types/const_test.go

```diff
@@ -153,16 +153,226 @@ func (s *testMySQLConstSuite) TestPipesAsConcatMode(c *C) {
 	r.Check(testkit.Rows("helloworld"))
 }
 
+func (s *testMySQLConstSuite) TestIssue22387(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	ctx := context.Background()
+	tk.MustExec("set sql_mode=''")
+	rs, _ := tk.Exec("select 12 - cast(15 as unsigned);")
```

Comment:
I tested in MySQL 5.7. No matter the value of sql_mode, this query will raise the same error.